File: python/leetcode/234. Palindrome Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        if head is None:
            return True

        late = head
        early = head.next
        prev = None

        while early:
            early = early.next.next if early.next else None
            late = late.next

        cur = None
        mid = late
        late = None

        while mid:
            cur = mid
            mid = mid.next
            cur.next = prev
            prev = cur

        while cur:
            if head.val != cur.val:
                return False
            head = head.next
            cur = cur.next

        return True

File: c++/leetcode/234. Palindrome Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution
{
public:
  bool isPalindrome(ListNode *head)
  {
    if (!head)
      return true;
    ListNode *late = head;
    ListNode *early = head->next;
    ListNode *prev = nullptr;

    while (early)
    {
      early = early->next ? early->next->next : nullptr;
      late = late->next;
    }

    // divide list
    ListNode *mid = late;
    late = nullptr;
    ListNode *cur = nullptr;

    // reverse list half
    while (mid)
    {
      cur = mid;
      mid = mid->next;
      cur->next = prev;
      prev = cur;
    }

    // compare values
    while (cur)
    {
      if (head->val != cur->val)
        return false;
      head = head->next;
      cur = cur->next;
    }

    return true;
  }
};